alg/los.cpp

```diff
@@ -0,0 +1,224 @@
+// * cmake --build build --target gdal_unit_test -j `nproc`
+// * ./build/autotest/cpp/gdal_unit_test --gtest_filter="test_alg.GDALIsLineOfSightVisible*"
+// Or, with GDB
+// *  gdb ./build/autotest/cpp/gdal_unit_test
+// * >>> r --gtest_filter="test_alg.GDALIsLineOfSightVisible*"
+
+#include <functional>
+#include <cmath>
+
+
+#include "cpl_port.h"
+#include "gdal_alg.h"
+
+// There's a plethora of bresenham implementations, all questionable production quality.
+// Bresenham optimizes for integer math, which makes sense for raster datasets in 2D.
+// For 3D, a 3D bresenham could be used if the altitude is also integer resolution.
+// 2D:
+// https://codereview.stackexchange.com/questions/77460/bresenhams-line-algorithm-optimization
+// https://gist.github.com/ssavi-ict/092501c69e2ffec65e96a8865470ad2f
+// https://blog.demofox.org/2015/01/17/bresenhams-drawing-algorithms/
+// https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm
+// https://www.cs.helsinki.fi/group/goa/mallinnus/lines/bresenh.html
+// https://stackoverflow.com/questions/10060046/drawing-lines-with-bresenhams-line-algorithm
+// http://www.edepot.com/linebresenham.html
+// 3D:
+// https://gist.github.com/yamamushi/5823518
+
+// Run bresenham terrain checking from (x1, y1) to (x2, y2).
+// The callback is run at every point along the line,
+// which should return True if the point is above terrain.
+// Bresenham2D will return true if all points have LOS between the start and end.
+bool Bresenham2D(int x1, int y1, int x2, int y2, std::function< auto(int, int) -> bool> OnBresenhamPoint) {
+    bool isAboveTerrain = true;
+	int	dx, dy;
+	int	incx, incy;
+
+	if (x2 >= x1)
+	{
+		dx = x2 - x1;
+		incx = 1;
+	}
+	else
+	{
+		dx = x1 - x2;
+		incx = -1;
+	}
+
+	if (y2 >= y1)
+	{
+		dy = y2 - y1;
+		incy = 1;
+	}
+	else
+	{
+		dy = y1 - y2;
+		incy = -1;
+	}
+
+	auto x = x1;
+	auto y = y1;
+    int	balance;
+
+	if (dx >= dy)
+	{
+		dy <<= 1;
+		balance = dy - dx;
+		dx <<= 1;
+
+		while (x != x2 && isAboveTerrain)
+		{
+			isAboveTerrain &= OnBresenhamPoint(x, y);
+			if (balance >= 0)
+			{
+				y += incy;
+				balance -= dx;
+			}
+			balance += dy;
+			x += incx;
+		} 
+        isAboveTerrain &= OnBresenhamPoint(x, y);
+
+	}
+	else
+	{
+		dx <<= 1;
+		balance = dx - dy;
+		dy <<= 1;
+
+		while (y != y2  && isAboveTerrain)
+		{
+			isAboveTerrain &= OnBresenhamPoint(x, y);
+			if (balance >= 0)
+			{
+				x += incx;
+				balance -= dy;
+			}
+			balance += dx;
+			y += incy;
+		}
+        isAboveTerrain &= OnBresenhamPoint(x, y);
+	}
+    return isAboveTerrain;
+}
+
+
+// Get the elevation of a single point.
+bool GetElevation(const GDALRasterBandH hBand, const int x, const int y, double& val)
+{
+    CPLAssert(x >= 0);
+    CPLAssert(x <= GDALGetRasterBandXSize(hBand));
+    CPLAssert(y >= 0);
+    CPLAssert(y <= GDALGetRasterBandYSize(hBand));
+
+    return GDALRasterIO(hBand, GF_Read, x, y, 1, 1, &val, 1, 1, GDT_Float64, 0, 0) == CE_None;
```

Comment:
I suspect using GDALCachedPixelAccessor (gcore/gdalcachedpixelaccessor.h ) would lead to better performance. Single pixel picking with GDALRasterIO() can have pretty big penalty. But could be done as a follow up pull request

Reply:
Thanks for the tip. I'll try it out if it's easy.

Reply:
For now, I added a TODO. I'd like to get some more representative test scenarios written up that I can use to benchmark it, but don't think it's necessary for the initial merge.